Add Temporary Tab

thecoolwinter opened the pull request:
# Description

This PR adds a temporary tab feature described in #604. When a new file is opened, it is opened as the temporary tab, if another file is opened right after the temporary tab is replaced with the new file.

To make a temporary tab permanent, the file can be double clicked, or the file can be edited.

- [x] Make a temporary tab when a new file is opened
- [x] Make a temporary tab permanent when it is double clicked
- [x] Make a temporary tab permanent when the file is edited
- [x] Remove animation when replacing temporary tab
- [x] Ensure both flavors of tabs work with the temporary tab
- [x] Add documentation to all the new methods in `WorkspaceDocument`
- [x] ~Make sure #601 still works as intended with this new feature~
- [x] Tab right-click option to keep open

# Related Issue

* #604 

# Checklist

- [x] I read and understood the [contributing guide](https://github.com/CodeEditApp/CodeEdit/blob/main/CONTRIBUTING.md) as well as the [code of conduct](https://github.com/CodeEditApp/CodeEdit/blob/main/CODE_OF_CONDUCT.md)
- [x] My changes generate no new warnings
- [x] My code builds and runs on my machine
- [x] I documented my code
- [x] Review requested

# Screenshots

https://user-images.githubusercontent.com/35942988/172027375-b3423977-3891-4fad-a20f-748f39b099e0.mov

https://user-images.githubusercontent.com/35942988/172027385-f11dbad5-c096-4228-81a8-84198021fdfd.mov

lilingxi01 commented:
#601 has some technical difficulties that are still WIP. If you are able to finish this PR not too long, I can handle the rest (e.g. potential conflicts).

thecoolwinter commented:
@lilingxi01 that sounds good. Right now I'm slightly stuck on listening for the file being saved but I think I can have this PR ready for review by tomorrow if all goes well.

austincondiff commented:
@thecoolwinter can you update the PR description with an updated video?

lilingxi01 commented on CodeEdit/Documents/WorkspaceDocument.swift:
```diff
@@ -157,6 +211,18 @@ final class WorkspaceDocument: NSDocument, ObservableObject, NSToolbarDelegate {
+    @objc func editorUpdatedContent() {
+        if selectionState.selectedId == selectionState.temporaryTab
+            && selectionState.temporaryTab != nil {
+            let tabID: TabBarItemID = selectionState.temporaryTab!
```
Please use `guard` instead of force unwrap.

thecoolwinter replied:
Got it!

lilingxi01 reviewed (changes requested):
The code looks good to me except one force unwrapping syntax.

I notice that this PR is enabling the address sanitizer in project scheme. After some [searches](https://developer.apple.com/documentation/xcode/diagnosing-memory-thread-and-crash-issues-early), I think it is for debugging purpose only and is not harmful to the project, so it is okay for me. If anyone has a different opinion on this, please share.

----

@austincondiff Just a note: in the future, we may need another PR to implement a missing feature and fix an animation bug:
* Double clicking a temporary tab should make it persist (like Xcode).
* When switching a tab from temporary to persist, there is a repeat animation (see video).

https://user-images.githubusercontent.com/36816148/172016709-c60d7989-35ee-4db3-a90e-3dd064fa14f7.mp4

wdg reviewed (dismissed):
Agreed, only point: please change the force-unwrap.

thecoolwinter commented:
Okay this is really ready!

In these last few commits I reworked a few things. First I made it so there isn’t a static “temporary tab” that’s always at the end of the list of tabs. Hopefully this makes #601 easier to work with these changes b/c it means the temporary tab acts just as a normal tab.

I also modified the `.onAppear` for `TabBarItem`s and made it so it determines by itself if it should animate. It uses a `previousTemporaryTab` variable on the selection state to determine this.

I also added a right-click option on the tab bar items for keeping the temp tab open as I noticed VSCode does this and it makes sense. That option only appears on temporary tabs.

Also, I moved the editing notification from the `textDidChange` to `textDidBeginEditing` to make it more efficient.

lilingxi01 commented:
@austincondiff : The testing CI of this PR is taking too long queuing (around 2 hours).

austincondiff commented:
@thecoolwinter Way to go on this PR! This makes CodeEdit feel much nicer to use.
